api/api.go

```diff
@@ -1,3 +1,114 @@
 // Copyright 2017, Square, Inc.
 
 package api
+
+import "github.com/square/etre/router"
+
+const (
+	API_ROOT              = "/api/v1/"
+	REQUEST_ID_PATTERN    = "([0-9]+)"
+	REQUEST_LABEL_PATTERN = "([A-Za-z0-9]+)"
+)
+
+// API provides controllers for endpoints it registers with a router.
+type API struct {
+	Router *router.Router
+}
+
+// NewAPI makes a new API.
+func NewAPI(router *router.Router) *API {
+	api := &API{
+		Router: router,
+	}
+
+	api.Router.AddRoute(API_ROOT+"entity/"+REQUEST_ID_PATTERN, api.entityHandler, "api-entity")
+	api.Router.AddRoute(API_ROOT+"entity/"+REQUEST_ID_PATTERN+"/labels", api.entityLabelsHandler, "api-entity-labels")
+	api.Router.AddRoute(API_ROOT+"entity/"+REQUEST_ID_PATTERN+"/labels"+REQUEST_LABEL_PATTERN, api.entityDeleteLabelHandler, "api-entity-delete-label")
+	api.Router.AddRoute(API_ROOT+"entities", api.entitiesHandler, "api-entities")
+	api.Router.AddRoute(API_ROOT+"query", api.queryHandler, "query-entity")
+	api.Router.AddRoute(API_ROOT+"stats", api.statsHandler, "api-stats")
+
+	return api
+}
+
+// ============================== CONTROLLERS ============================== //
+
+// {POST,GET,PUT,DELETE} /entity/{_id}
+// Managing a single entity
+func (api *API) entityHandler(ctx router.HTTPContext) {
+	switch ctx.Request.Method {
```

Comment:
Never mind: our bespoke router doesn't support per-method routes, so until we use another router, `switch` is fine, just put per-method handler code in separate functions and call those functions from the `switch` cases.